oauthproxy_test.go

```diff
@@ -1776,10 +1780,75 @@ func TestGetRedirect(t *testing.T) {
 			url:              proxy.ProxyPrefix + "/foo/bar",
 			expectedRedirect: "/",
 		},
+		{
+			name: "proxied request outside of ProxyPrefix redirects to proxiexd URL",
+			url:  "https://oauth.example.com/foo/bar",
+			headers: map[string]string{
+				"X-Forwarded-Proto": "https",
+				"X-Forwarded-Host":  "a-service.example.com",
+				"X-Forwarded-Uri":   "/foo/bar",
+			},
+			expectedRedirect: "https://a-service.example.com/foo/bar",
+		},
+		{
+			name: "proxied request with query preserves query",
+			url:  "https://oauth.example.com/foo?bar",
+			headers: map[string]string{
+				"X-Forwarded-Proto": "https",
+				"X-Forwarded-Host":  "a-service.example.com",
+				"X-Forwarded-Uri":   "/foo?bar",
+			},
+			expectedRedirect: "https://a-service.example.com/foo?bar",
+		},
+		{
+			name: "proxied request under ProxyPrefix redirects to root",
+			url:  "https://oauth.example.com" + proxy.ProxyPrefix + "/foo/bar",
+			headers: map[string]string{
+				"X-Forwarded-Proto": "https",
+				"X-Forwarded-Host":  "a-service.example.com",
+				"X-Forwarded-Uri":   proxy.ProxyPrefix + "/foo/bar",
+			},
+			expectedRedirect: "https://a-service.example.com/",
+		},
+		{
+			name: "proxied request with port under ProxyPrefix redirects to root",
+			url:  "https://oauth.example.com" + proxy.ProxyPrefix + "/foo/bar",
+			headers: map[string]string{
+				"X-Forwarded-Proto": "https",
+				"X-Forwarded-Host":  "a-service.example.com:8443",
+				"X-Forwarded-Uri":   proxy.ProxyPrefix + "/foo/bar",
+			},
+			expectedRedirect: "https://a-service.example.com:8443/",
+		},
+		{
+			name: "proxied request with missing uri header would still redirect to desired redirect",
+			url:  "https://oauth.example.com/foo?bar",
+			headers: map[string]string{
+				"X-Forwarded-Proto": "https",
+				"X-Forwarded-Host":  "a-service.example.com",
+			},
+			expectedRedirect: "https://a-service.example.com/foo?bar",
+		},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
 			req, _ := http.NewRequest("GET", tt.url, nil)
+			for header, value := range tt.headers {
+				req.Header.Add(header, value)
+			}
+			redirect, err := proxy.GetRedirect(req)
+
+			assert.NoError(t, err)
+			assert.Equal(t, tt.expectedRedirect, redirect)
+		})
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			req, _ := http.NewRequest("GET", tt.url, nil)
+			err = req.ParseForm()
+			assert.NoError(t, err)
+			req.Form.Add("rd", tt.expectedRedirect)
```

Comment:
We should be able to handle this in handle this in 1 subtest loop for the table (this will result in 2 sets of subtests with the same name).

I think for the `rd` cases we want, you can just add some test cases with `url` that has `?rd=<some other redirect URL>`

The underlying `GetRedirect` we are testing calls `req.ParseForm` immediately on the `http.Request` which will handle parsing the querystring `rd` in the URL into the `req.Form`.  So it will be good to let that happen in the test case rather than explicitly adding it ourselves here.